File: src/pkg/schemas/discovery.py
from datetime import datetime

from pydantic import BaseModel, Field

from pkg.schemas.source import SourceRead


class DiscoveryItemRead(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    user_id: str
    provider: str
    item_key: str
    title: str
    url: str | None = None
    summary: str | None = None
    payload: dict
    status: str
    score: float | None = None
    why: list | None = None
    source_id: str | None = None
    feedback: dict | None = None
    created_at: datetime
    updated_at: datetime
    reviewed_at: datetime | None = None


class DiscoveryItemList(BaseModel):
    items: list[DiscoveryItemRead]
    total: int


class DiscoveryGenerateRequest(BaseModel):
    providers: list[str] = Field(default_factory=lambda: ["github", "rss", "web"])
    limit: int = Field(default=50, ge=1, le=200)


class DiscoveryGenerateResult(BaseModel):
    created: int
    updated: int
    skipped: int


class DiscoveryWebResult(BaseModel):
    title: str
    url: str
    summary: str | None = None
    source_name: str | None = None
    published_at: str | None = None


class DiscoveryWebIngestRequest(BaseModel):
    query: str
    items: list[DiscoveryWebResult] = Field(default_factory=list, min_length=1, max_length=50)


class DiscoveryWebSearchRequest(BaseModel):
    query: str
    max_results: int = Field(default=10, ge=1, le=20)


class DiscoveryFeedbackRequest(BaseModel):
    action: str = Field(pattern=r"^(keep|save|dismiss)$")
    note: str | None = None


class DiscoveryFeedbackResult(BaseModel):
    item: DiscoveryItemRead
    source: SourceRead | None = None
    created: bool | None = None

File: web/src/lib/api/discovery.ts
import { request } from "./client";
import type { Source } from "./sources";

export type DiscoveryStatus = "recommended" | "kept" | "saved" | "dismissed";
export type DiscoveryProvider = "arxiv" | "github" | "rss" | "web" | "openalex" | "crossref" | "semantic_scholar";

export interface DiscoveryItem {
  id: number;
  user_id: string;
  provider: DiscoveryProvider;
  item_key: string;
  title: string;
  url?: string | null;
  summary?: string | null;
  payload: Record<string, unknown>;
  status: DiscoveryStatus;
  score?: number | null;
  why?: string[] | null;
  source_id?: string | null;
  feedback?: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  reviewed_at?: string | null;
}

export interface DiscoveryItemList {
  items: DiscoveryItem[];
  total: number;
}

export interface DiscoveryWebResult {
  title: string;
  url: string;
  summary?: string | null;
  source_name?: string | null;
  published_at?: string | null;
}

export const discoveryApi = {
  list: (params?: { provider?: DiscoveryProvider; status?: DiscoveryStatus; limit?: number; offset?: number }) => {
    const q = new URLSearchParams();
    if (params?.provider) q.set("provider", params.provider);
    if (params?.status) q.set("status", params.status);
    if (params?.limit) q.set("limit", String(params.limit));
    if (params?.offset) q.set("offset", String(params.offset));
    return request<DiscoveryItemList>(`/discovery?${q}`);
  },
  generate: (body?: { providers?: DiscoveryProvider[]; limit?: number }) =>
    request<{ created: number; updated: number; skipped: number }>("/discovery/generate", {
      method: "POST",
      body: JSON.stringify(body ?? {}),
    }),
  ingestWebResults: (body: { query: string; items: DiscoveryWebResult[] }) =>
    request<{ created: number; updated: number; skipped: number }>("/discovery/web-results", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  searchWeb: (body: { query: string; max_results?: number }) =>
    request<{ created: number; updated: number; skipped: number }>("/discovery/web-search", {
      method: "POST",
      body: JSON.stringify(body),
    }),
  feedback: (id: number, action: "keep" | "save" | "dismiss", note?: string | null) =>
    request<{ item: DiscoveryItem; source?: Source | null; created?: boolean | null }>(`/discovery/${id}/feedback`, {
      method: "PATCH",
      body: JSON.stringify({ action, note }),
    }),
};
